### DISCUSSION TITLE
Three JS integration?

### DISCUSSION THREAD

**Discussion by erikpa1** [tauri-apps/tauri#5045] [Q&A]
Hi, 

I have assets viewer (3d meshes with json data), it is static app written in Three JS deployed on azure storage, which can work as static web app with 5TB capacity

Files structure is 

-mytauriapp.exe
--assets
---myAsset1/files...
---myAsset2/files...

But now, I want to create desktop app but I don't have control on files fetch functinality (its blackbox in three js for me)

How can I create app like this? I can't rewrite every three js library to use '@tauri-apps/api/fs'

Is there any type of syntax to access file via fetch  like   fetch("$APP/../assets/myAsset/myMesh.glb") ???

Or what are ways to solve this use case?

**Comment by vdvman1** [comment]
You can use the [`convertFileSrc`](https://tauri.app/v1/api/js/modules/tauri#convertfilesrc) method to get an asset URL for a given path within the app.
The asset protocol looks different on different platforms, which is why it is a method

**Reply by erikpa1** [reply]
Thanks, I will try

**Reply by erikpa1** [reply]
@vdvman1  Update, not working three JS cant find this file, I can create folder in that directory but not fetch it. 


File get function:
```
    static async convertFileResourcePath(path) {
        const joinedPath = await join(await resourceDir(), `resources/${path}`)
        const assetUrl = convertFileSrc(joinedPath);
        return assetUrl

    }
```

React fiber code:

```
const AsyncMaterial = ({path}) => {

    console.log("Heere")

    const texture = useLoader(three.TextureLoader, path)

    return (
        <meshBasicMaterial transparent={true} map={texture} side={three.DoubleSide}/>
    )
}


export const World = ({}) => {

    const [convertedPath, setConvertedPath] = React.useState("")

    React.useEffect(() => {
        StorageManager.convertFileResourcePath("tmp.png").then((path) => {
            setConvertedPath(path)
            // console.log(path)
        })
    }, [])

    console.log(convertedPath !== "")

    return (
        <Canvas>
            <ambientLight/>

            <pointLight position={[10, 10, 10]}/>
            <Box position={[-1.2, 0, 0]}>
                {
                    convertedPath !== "" && <AsyncMaterial path={convertedPath}/>
                }
            </Box>
            <Box position={[1.2, 0, 0]}/>
        </Canvas>
    )
}

```

File is where should be:

![image](https://user-images.githubusercontent.com/19526670/187047180-6bd6ada3-60e0-4dda-8e99-a9f21c9c3743.png)

![image](https://user-images.githubusercontent.com/19526670/187047148-bcdae6eb-6e87-4aba-87fb-2a6fa2c62742.png)


```
Uncaught Error: Could not load https://asset.localhost/%5C%5C%3F%5CC%3A%5CWork%5Cexedevelopment%5Cinfinity%5Cinfinity_rust%5Csrc-tauri%5Ctarget%5Cdebug%5Cresources%2Ftmp.png: undefined)
    at index-cf0284c7.esm.js:1616:1
    at HTMLImageElement.onImageError (three.module.js:40234:1)
```

**Reply by JonasKruckenberg** [reply]
The asset protocol won't help you in this case, since its job is to serve the files that are inline into the binary., not any file on disk.

I would recommend you check out the Custom Protocol API we have and define your own url scheme like `model://<relative path to model>`

https://docs.rs/tauri/1.0.5/tauri/struct.Builder.html#method.register_uri_scheme_protocol

You can see [this example](https://github.com/tauri-apps/tauri/blob/dev/examples/streaming/main.rs) for how to use the custom uri scheme

**Reply by erikpa1** [reply]
@JonasKruckenberg I'm just rust and tauri beginner, so don't exactly know what you are sending here, but don't you think its really big overkill for so simple and basic think I'm doing?  

Tauri really don't have some config setting which make one of the folder public for any file developers want to have ???

**Reply by vdvman1** [reply]
Having Tauri automatically be able to load files from disk is a security risk, since it could end up running malicious JS files in your webview.
If it were added, it would need to be audited, and very carefully designed (likely using some config with a scope in the allowlist).
So it is not something that can easily be officially added to Tauri itself, but it is something you can add yourself to your app relatively easily, as long as you are willing to accept the potential security risks

**Reply by JonasKruckenberg** [reply]
Actually turns out I'm just stupid, the asset protocol can indeed load any file on disk given you have configured your scope accordingly: https://tauri.app/v1/guides/development/security#asset-protocol

Your issue is that you're trying to load the file from the wrong directory, `resourceDir` is for resources that a bundled alongside your application (and I assume you don't want to bundle 20GB files with your app). 

What you want instead is `appDir()` for user files.

**Reply by erikpa1** [reply]
@JonasKruckenberg Yes but if I remember right appdir   wasn't near exe file it was in `C:\Users\erikp\AppData` which is not something  I want, I want these files to have near exe file.  

My use case is:

I have c++ desktop app which generates webapp with panorama projects but I want to give same player (tauri in this case) for my customer as desktop version. 

My desktop app generates 

- player.exe
- panorama
- ---projectA
- ---projectB
- .....

User takes USB drive, directory copy files from his PC and takes them to another PC.

**Reply by erikpa1** [reply]
@vdvman1 

No its not security risk (for my kind of application), because source code is still hidden inside  .exe file I'm not loading JS files outside of .exe file, I'm loading just photos videos and json config files, no code. My app don't need any fs or other permissions  only access to this one folder.

**Reply by erikpa1** [reply]
@JonasKruckenberg Tried asset protocol, still not working for three JS texture loader (fetch for file works and` <img src={convertedPath}/>` is also is working),

Not sure what is happening because three js uses   "img"  element which handles loading :D :D 

```
    static async convertFileResourcePath(path) {
        //await resourceDir()

        const appDirPath = await appDir()

        const joinedPath = await join(appDirPath, "assets", path)
        const assetUrl = convertFileSrc(joinedPath);
        return assetUrl

    }
```

![image](https://user-images.githubusercontent.com/19526670/187357935-a5b73fa6-0f65-4186-be7f-59dfc508e2c7.png)

Tauri config:

```
  "tauri": {
    "allowlist": {
      "all": true,
      "fs": {
        "all": true
      },
      "path": {
        "all": true
      },
      "protocol": {
        "asset": true,
        "assetScope": [
          "$APP/assets/*"
        ]
      }
    },

```

**Reply by JonasKruckenberg** [reply]
Have you tried to bypass threejs for debugging purposes and render an `<img>` tag with the url directly? Maybe you can start to isolate wether it's a problem with tauri or somewhere in between.

**Reply by erikpa1** [reply]
@JonasKruckenberg  img tag is working as written above, also fetch() works fine. I also asked on three JS forum but still no answer :) when I will come home from job I will cut piece off code and put it on public git where you can try if you have time. Because now its on private devops repository :)

**Reply by erikpa1** [reply]
@JonasKruckenberg  https://github.com/erikpa1/tauri_threejs  here is repository I created, should work on first time after writing command I have typed to read.me, thank you for your time :) there is `<World>` component commented in `<App> `component uncoment it and it should crash :)

**Reply by erikpa1** [reply]
@JonasKruckenberg UPDATE, today project magically works, not sure if there was some problem with cache or something but project works today:

![image](https://user-images.githubusercontent.com/19526670/187988777-734cd791-28be-4367-a8bd-461c4d0786a2.png)
